pulser/simulation/simulation.py

```diff
@@ -325,55 +463,125 @@ def get_hamiltonian(self, time: float) -> qutip.Qobj:
                              "greater than or equal to 0.")
         return self._hamiltonian(time/1000)  # Creates new Qutip.Qobj
 
+    def _reset_samples(self) -> None:
+        """Sets all samples to their default values.
+
+        Used when running several noisy runs, each one setting new samples.
+        """
+        self.samples = {addr: {basis: {}
+                               for basis in ['ground-rydberg', 'digital']}
+                        for addr in ['Global', 'Local']}
+
+    def _init_dephasing(self) -> None:
+        """Initializes dephasing collapse operators."""
+        if self.basis_name == 'digital' or self.basis_name == 'all':
+            raise ValueError("Cannot include dephasing noise in digital-"
+                             "or all-basis.")
+        self.dephasing_prob = 0.05  # Probability of phase (Z) flip
+        self._collapse_ops += [
+            np.sqrt(1 - self.dephasing_prob) * self.op_matrix['I'],
+            np.sqrt(self.dephasing_prob) * (self.op_matrix['sigma_rr']
+                                            - self.op_matrix['sigma_gg'])]
+
+    def _build_hamiltonian_from_seq(self) -> None:
+        """Builds the hamiltonian from sequence samples.
+
+        Updates the new sequence samples that may have changed due to noise,
+        builds default operators and builds the hamiltonian from those
+        coefficients and operators.
+        """
+        self._reset_samples()
+        self._extract_samples()
+        self._build_basis_and_op_matrices()
```

Comment:
I left a comment with a change that might justify doing this every time, but for now there's no reason to repeat this call more than once, right?

Reply:
Maybe for the last method, but for the other two, samples change every run according to random parameters due to noise, so we do need to `extract_samples`every run.

Reply:
Yes, I meant `_build_basis_and_op_matrices` in particular.